Reflection_Engine/Compute/RunExtensionMethod.cs

```diff
@@ -36,72 +38,131 @@ public static partial class Compute
+                return GetStoredExtensionMethod(key)?.Invoke(parameters);
+
+            //Loop through all methods with matching name, first argument and number of parameters, sorted by best match to the first argument
+            foreach (MethodInfo method in type.ExtensionMethods(methodName).Where(x => x.GetParameters().Length == parameters.Length).SortExtensionMethods(type))
             {
                 ParameterInfo[] paramInfo = method.GetParameters();
 
-                // Make sure the type of parameters is matching
+                // Make sure the type of parameters is matching, skipping first as already used to extract parameters
                 bool matchingTypes = true;
-                for (int i = 0; i < parameters.Length; i++)
+                for (int i = 1; i < parameters.Length; i++)
                 {
-                    if (!paramInfo[i + 1].ParameterType.IsAssignableFrom(parameters[i].GetType()))
+                    if(!paramInfo[i].ParameterType.IsAssignableFromIncludeGenerics(parameters[i].GetType()))
                     {
+                        //Parameter does not match, abort for this method
                         matchingTypes = false;
                         break;
                     }
                 }
                 if (!matchingTypes)
                     continue;
 
-                m_PreviousInvokedMethods[key] = method;
-                return method.Invoke(null, new object[] { target }.Concat(parameters).ToArray());
+                MethodInfo finalMethod = method;
+
+                //If method is generic, make sure the appropriate generic arguments are set
+                if (method.IsGenericMethod)
+                    finalMethod = method.MakeGenericFromInputs(parameters.Select(x => x.GetType()).ToList());
+
+                //TUrn the MethodInfo to a compiled function, store it and finally call it
+                Func<object[], object> func = finalMethod.ToFunc();
+                StoreExtensionMethod(key, func);
+                return func(parameters);
             }
 
+            //If nothing found, store null, to avoid having to search again in vain
+            StoreExtensionMethod(key, null);
+
             // Return null if nothing found
             return null;
         }
 
+        /***************************************************/
+
+        [Description("Checkes if an entry with the provided key has already been extracted. Put in its own method to simplify the use of locks to provide thread safety.")]
+        private static bool MethodPreviouslyExtracted(Tuple<Type, string> key)
+        {
+            lock (m_RunExtensionMethodLock)
+            {
+                return m_PreviousInvokedMethods.ContainsKey(key);
+            }
+        }
+
+        /***************************************************/
+
+        [Description("Gets a previously extracted method from the stored methods. Put in its own method to simplify the use of locks to provide thread safety.")]
```

Comment:
Reason I wanted to keep both was that I am storing nulls for cases where no method is found. This is to avoid searching again and again when no method exist for a particular set of inputs. Say for example displaying geometry, it is being called for all BHoMObejct types. Without storing null, it will keep on searching again and again for an applicable method, all in vain.

This means "null" is a valid answer for "ContainsKey", hence can't use that to determaine if the method has previously been attempted to be extracted or not.

TL;DR

`MethodPreviouslyExtracted` checks if an attempt has been made to extract the method.
`GetStoredExtensionMethod` returns is if existing. Can be null for a case when an attempt has been made before, but failed to find anything -> No point trying again. (nulls are filtered out by the maybe not so visible `?` sign on the line it is being called)

Reply:
Right, makes perfect sense.